src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -106,6 +106,93 @@ protected int getCacheLimit() {
+                "isEmpty(second) == isSecondEmpty",
+                "getCodeRange(first) == firstCodeRange",
+                "getCodeRange(second) == secondCodeRange",
+                "getEncoding(first) == firstEncoding",
+                "getEncoding(second) == secondEncoding"
+        }, limit = "getCacheLimit()")
+        protected Encoding negotiateRopeRopeCached(Rope first, Rope second,
+                @Cached("getEncoding(first)") Encoding firstEncoding,
+                @Cached("getEncoding(second)") Encoding secondEncoding,
+                @Cached("isEmpty(first)") boolean isFirstEmpty,
+                @Cached("isEmpty(second)") boolean isSecondEmpty,
+                @Cached("getCodeRange(first)") CodeRange firstCodeRange,
+                @Cached("getCodeRange(second)") CodeRange secondCodeRange,
+                @Cached("negotiateRopeRopeUncached(first, second)") Encoding negotiatedEncoding) {
+            return negotiatedEncoding;
+        }
+
+        @Specialization(guards = "getEncoding(first) != getEncoding(second)", replaces = "negotiateRopeRopeCached")
+        protected Encoding negotiateRopeRopeUncached(Rope first, Rope second) {
+            return compatibleEncodingForRopes(first, second);
+        }
+
+        protected Encoding getEncoding(Rope rope) {
+            return rope.getEncoding();
+        }
+
+        protected CodeRange getCodeRange(Rope rope) {
+            return rope.getCodeRange();
+        }
+
+        protected boolean isEmpty(Rope rope) {
+            return rope.isEmpty();
+        }
+
+        @TruffleBoundary
+        private static Encoding compatibleEncodingForRopes(Rope firstRope, Rope secondRope) {
+            // Taken from org.jruby.RubyEncoding#areCompatible.
+
+            final Encoding firstEncoding = firstRope.getEncoding();
+            final Encoding secondEncoding = secondRope.getEncoding();
+
+            if (secondRope.isEmpty())
+                return firstEncoding;
+            if (firstRope.isEmpty()) {
+                return firstEncoding.isAsciiCompatible() && (secondRope.getCodeRange() == CodeRange.CR_7BIT) ? firstEncoding : secondEncoding;
+            }
+
+            if (!firstEncoding.isAsciiCompatible() || !secondEncoding.isAsciiCompatible())
+                return null;
+
+            if (firstRope.getCodeRange() != secondRope.getCodeRange()) {
+                if (firstRope.getCodeRange() == CodeRange.CR_7BIT)
+                    return secondEncoding;
+                if (secondRope.getCodeRange() == CodeRange.CR_7BIT)
+                    return firstEncoding;
+            }
+            if (secondRope.getCodeRange() == CodeRange.CR_7BIT)
+                return firstEncoding;
+            if (firstRope.getCodeRange() == CodeRange.CR_7BIT)
+                return secondEncoding;
+
+            return null;
+        }
+
+        protected int getCacheLimit() {
+            return getContext().getOptions().ENCODING_COMPATIBLE_QUERY_CACHE;
+        }
+
+    }
```

Comment:
Can the node below reuse this logic then?

Reply:
I've been considering that. The node below is quite complex, so I'd like to do the change separately so that any changes in behaviour aren't too hard to track down.

Reply:
Should the `compatibleEncodingForRopes` be a node?

Reply:
Yes, but I think we really want to restructure quite a few of the specialisations. Most of the time we want something like
```
        @Specialization(guards = {
                "isRubyString(first)",
                "isRubyString(second)" })
        protected Encoding negotiateSameEncodingCached(DynamicObject first, DynamicObject second,
                                                       @Cached("...") NegotiateCompatibleRopeEncodingNode ropeNode) {
            return ropeNode.execute(first, second);
        }
```
and then only deal with the odd cases in NegotiateCompatibleEncodingNode.

Reply:
> I've been considering that. The node below is quite complex, so I'd like to do the change separately so that any changes in behaviour aren't too hard to track down.

That's actually the reason why I don't want any duplication. If the behavior of one of these two duplicates changes and break something, it's gonna be very strange to track down. So let's deduplicate and if it's changed then it has to stay compatible will all usages (or the usages be adapted).

Reply:
I changed this to use the rope node for all strings, and removed the complex duplicate logic.